pkg/plugins/client.go

```diff
@@ -132,8 +132,13 @@ func (c *Client) callWithRetry(serviceMethod string, data io.Reader, retry bool)
 	var retries int
 	start := time.Now()
 
+	reqBody, err := ioutil.ReadAll(data)
```

Comment:
Unfortunately this won't work because this stream can be huge and eat up all the memory.

Reply:
You're right, I missed the GraphDriver plugin code path.

What do you suggest here? I see multiple options:
1) A `TeeReader` that would rebuild the body as it is consumed, in case we need to retry the request (unlikely, so wasteful).
2) Try to convert the API to use `io.ReadSeeker`, probably only marginally better than `ioutil.ReadAll`.
3) Do not attempt to retry if the connection was established with the plugin server, since bytes might have been read from the body. It would simply require checking the error type.

Reply:
4. Do not use `callWithRetry` in `SendFile`, since it might be the only case where memory can blow up. Instead, switch `SendFile` to another function that do no retry. Other plugin API methods will still retry.

Reply:
In this case I think it's fair to return an error rather than retry and let the caller deal with retrying. We can't handle this generically and correctly enough this deep in the stack.
We could ask for a seeker, but this is pretty limiting and may not be possible to retry in all cases.

Reply:
> In this case I think it's fair to return an error rather than retry and let the caller deal with retrying. We can't handle this generically and correctly enough this deep in the stack.

For the `Stream` function I agree with you and that returning an error seems like the correct behavior.

For the `Call(serviceMethod string, args interface{}, ret interface{})` function it seems to me that this PR would fix that bug.

Solution 4 seems to at least fix part of the problem :)

Reply:
The retry for this particular case could be handled in the `Call()` function.

Reply:
@cpuguy83 Are you saying we should unconditionally remove the retry logic for `Stream`/`SendFile`? I think that would be a regression in the case where it's the `Dial` step that fails, for instance if the plugin server is being initialized and the socket has not been created yet.

Therefore, the ideal would be:
- For `Call`: retry if `Dial` fails or if the connection is reset.
- For `SendFile`/`Stream`: retry only if `Dial` fails (since if the connection is reset, we can't rewind)
That's becoming a bit more tricky to handle, but do you agree with the above?

Reply:
Maybe we could just stop calling `callWithRetry` for requests with huge body, or we can provide a different interface with seekable stream like `callWithRetrySeekable` if it is really necessary. It is not necessary to solve the problem with exactly one interface.

Reply:
@flx42 I'm saying let `callWithRetry` deal with dial issues, and let the caller determine how to handle other issues. The real intention of `callWithRetry` is specifically for plugins that aren't ready as opposed to some other issue (like a dropped connection).

`Call` could seek the buffer back to the beginning and retry in certain cases. However, thinking about it determining if another attempt should be made after a partial transfer really seems to be the responsibility of the subsystem making the request rather than the low-level framework which doesn't really know how the plugin is expected to operate.

Reply:
What is the situation now? @flx42 @cpuguy83